### PR TITLE
Support "Load more" on runs table

### PR DESCRIPTION
**Pull request by Zangr**
## What changes are proposed in this pull request?
 
- Provide a "Load more" bar on the bottom of the runs table.
- Convert sort dropdown to be based on AntD which injects menu to HTML body to avoid it being covered by the load-more bar.

load-more-1

## How is this patch tested?
 Unit tests, Manual tests

(Details)
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
Provide a "Load more" bar on the bottom of the runs table.
 
### What component(s) does this PR affect?
 
- [x] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [ ] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/server/js/src/Actions.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import { MlflowService } from './sdk/MlflowService';
 import ErrorCodes from './sdk/ErrorCodes';
 
-export const SEARCH_MAX_RESULTS = 1000;
+export const SEARCH_MAX_RESULTS = 100;
```
Comment (mparkhe) [inline mlflow/server/js/src/Actions.js]: 👍

File: mlflow/server/js/src/components/BaggedCell.js [hunk 2]
```diff
@@ -24,56 +22,65 @@ export default class BaggedCell extends PureComponent {
...
+          overlay={(
+            <Menu>
+              <Menu.Item onClick={this.handleSortAscending}>
+                Sort ascending
+              </Menu.Item>
+              <Menu.Item onClick={this.handleSortDescending}>
+                Sort descending
+              </Menu.Item>
+              <Menu.Item onClick={this.handleRemoveBagged}>
+                Display in own column
```
Comment (mparkhe) [inline mlflow/server/js/src/components/BaggedCell.js]: Do we want to reword this to something like "Display as a separate column" or something like that?

cc: @stbof

File: mlflow/server/js/src/components/ExperimentView.js [hunk 3]
```diff
@@ -403,10 +400,7 @@ export class ExperimentView extends Component {
           </form>
           <div className="ExperimentView-run-buttons">
             <span className="run-count">
-              {runInfos.length > SEARCH_MAX_RESULTS ?
-                `Showing the latest ${SEARCH_MAX_RESULTS} matching runs` :
-                `${runInfos.length} matching ${runInfos.length === 1 ? 'run' : 'runs'}`
-              }
+              Showing the latest {runInfos.length} matching {runInfos.length === 1 ? 'run' : 'runs'}
```
Comment (mparkhe) [inline mlflow/server/js/src/components/ExperimentView.js]: Can we change this to something like 

> Showing ### matching runs

Latest made sense when the UI's `order_by` was `start_time DESC` (and no way to change it)

**Comment by gioa** [comment]
@Zangr For the load more button, can we use the existing button colors to be consistent? 
Also, could you make the button padding-left and padding-right 16px so it has more clicking area and looks thinner inside the table border?

File: mlflow/server/js/src/components/ExperimentPage.js [hunk 4]
```diff
@@ -31,75 +34,112 @@ export class ExperimentPage extends Component {
...
     if (props.experimentId !== state.lastExperimentId) {
-      const newState = {
+      return {
         ...ExperimentPage.getDefaultUnpersistedState(),
         persistedState: state.lastExperimentId === undefined ?
             state.persistedState : (new ExperimentPagePersistedState()).toJSON(),
         lastExperimentId: props.experimentId,
         lifecycleFilter: LIFECYCLE_FILTER.ACTIVE,
       };
-      props.getExperimentApi(props.experimentId, newState.getExperimentRequestId);
```
Comment (Zangr) [inline mlflow/server/js/src/components/ExperimentPage.js]: Generally, we should avoid sending requests in `getDerivedStateFromProps`. 
https://github.com/facebook/react/issues/13541#issuecomment-418101478

File: mlflow/server/js/src/components/ExperimentRunsTableCompactView.js [hunk 5]
```diff
@@ -485,10 +505,11 @@ class ExperimentRunsTableCompactView extends PureComponent {
                   }}
                   style={{...styles.columnStyle, borderLeft: BORDER_STYLE}}
                   cellRenderer={({rowIndex, rowData, parent, dataKey}) => {
-                    // Add extra padding to last row so that we can render dropdowns for bagged
-                    // param key-value pairs in that row
-                    const paddingOpt = rowIndex === rows.length - 1 ? {paddingBottom: 95} : {};
                     const colIdx = NUM_RUN_METADATA_COLS + unbaggedParams.length;
+                    // Add extra padding for load more
+                    const paddingOpt = rowIndex === rows.length - 1
+                      ? { paddingBottom: LOAD_MORE_ROW_HEIGHT * 2 }
```
Comment (Zangr) [inline mlflow/server/js/src/components/ExperimentRunsTableCompactView.js]: Ideally, this should equal to `FINAL_ROW_HEIGHT + LOAD_MORE_ROW_HEIGHT` with empty cell present in a row. But `FINAL_ROW_HEIGHT` is not available at this stage so I made this 2x the LOAD_MORE_ROW_HEIGHT as a rough accommodation for empty cells.

File: mlflow/server/js/src/components/ExperimentView.css [hunk 6]
```diff
@@ -232,7 +232,11 @@ span.error-message {
   font-weight: inherit;
 }
 
-/* Workaround as per https://github.com/bvaughn/react-virtualized/issues/876#issuecomment-367029529 */
+.ExperimentView .ReactVirtualized__Table__Grid:focus {
+  outline: 0;
```
Comment (Zangr) [inline mlflow/server/js/src/components/ExperimentView.css]: This is to remove the annoying blue frame when we click the table.

File: mlflow/server/js/src/reducers/Reducers.js [hunk 7]
```diff
@@ -49,21 +57,24 @@ export const getRunInfo = (runUuid, state) => {
...
 const runInfosByUuid = (state = {}, action) => {
   switch (action.type) {
-    case fulfilled(GET_EXPERIMENT_API): {
-      let newState = { ...state };
+    case fulfilled(GET_RUN_API): {
+      const runInfo = RunInfo.fromJs(action.payload.run.info);
+      return amendRunInfosByUuid(state, runInfo);
+    }
+    case fulfilled(SEARCH_RUNS_API): {
+      const newState = {};
```
Comment (Zangr) [inline mlflow/server/js/src/reducers/Reducers.js]: `runInfosByUuid` will now be the source of truth for runs to display now. Old runs will be thrown away after we call search.

File: mlflow/server/js/src/reducers/Reducers.js [hunk 8]
```diff
@@ -49,21 +57,24 @@ export const getRunInfo = (runUuid, state) => {
 
 const runInfosByUuid = (state = {}, action) => {
   switch (action.type) {
-    case fulfilled(GET_EXPERIMENT_API): {
```
Comment (Zangr) [inline mlflow/server/js/src/reducers/Reducers.js]: Ignore runs from `getExperiment`

File: mlflow/server/js/src/reducers/Reducers.js [hunk 9]
```diff
@@ -270,7 +283,7 @@ export const getSharedMetricKeysByRunUuids = (runUuids, state) =>
 
 export const getApis = (requestIds, state) => {
   return requestIds.map((id) => (
-    state.apis[id]
+    state.apis[id] || {}
```
Comment (Zangr) [inline mlflow/server/js/src/reducers/Reducers.js]: As we are moving fetchings out of `getDerivedStateFromProps`, this will address the NPE when we do not load before the first rendering.

File: mlflow/server/js/src/components/BaggedCell.js [hunk 10]
```diff
@@ -24,56 +22,65 @@ export default class BaggedCell extends PureComponent {
...
+          overlay={(
+            <Menu>
+              <Menu.Item onClick={this.handleSortAscending}>
+                Sort ascending
+              </Menu.Item>
+              <Menu.Item onClick={this.handleSortDescending}>
+                Sort descending
+              </Menu.Item>
+              <Menu.Item onClick={this.handleRemoveBagged}>
+                Display as a separate column
```
Comment (Zangr) [inline mlflow/server/js/src/components/BaggedCell.js]: This line changes `Display in own column` to `Display as a separate column`. @stbof

File: mlflow/server/js/src/components/BaggedCell.js (hunk 2 above)
Comment (Zangr) [reply mlflow/server/js/src/components/BaggedCell.js]: reworded

File: mlflow/server/js/src/components/ExperimentView.js (hunk 3 above)
Comment (Zangr) [reply mlflow/server/js/src/components/ExperimentView.js]: Removed `latest`

File: mlflow/server/js/src/components/ExperimentPage.js [hunk 11]
```diff
@@ -31,75 +34,112 @@ export class ExperimentPage extends Component {
...
+    const { experimentId } = this.props;
+    const { orderByKey, orderByAsc, searchInput } = persistedState;
+    const orderBy = ExperimentPage.getOrderByExpr(orderByKey, orderByAsc);
+    const viewType = lifecycleFilterToRunViewType(lifecycleFilter);
+
+    this.props.getExperimentApi(experimentId, this.getExperimentRequestId);
+    this.props
+      .searchRunsApi([experimentId], searchInput, viewType, orderBy, this.searchRunsRequestId)
+      .then(this.updateNextPageToken)
+      .catch(Utils.logErrorAndNotifyUser);
```
Comment (smurching) [inline mlflow/server/js/src/components/ExperimentPage.js]: Can we also unset the loadingMore state in this case? The `Utils.logErrorAndNotifyUser` handler seems to work well, it's just that the loadingMore spinner continues to spin because we don't do `this.setState({ nextPageToken, loadingMore: false });`

File: mlflow/server/js/src/components/ExperimentView.js [hunk 12]
```diff
@@ -403,10 +399,7 @@ export class ExperimentView extends Component {
           </form>
           <div className="ExperimentView-run-buttons">
             <span className="run-count">
-              {runInfos.length > SEARCH_MAX_RESULTS ?
-                `Showing the latest ${SEARCH_MAX_RESULTS} matching runs` :
-                `${runInfos.length} matching ${runInfos.length === 1 ? 'run' : 'runs'}`
-              }
+              Showing the {runInfos.length} matching {runInfos.length === 1 ? 'run' : 'runs'}
```
Comment (smurching) [inline mlflow/server/js/src/components/ExperimentView.js]:
```suggestion
              Showing {runInfos.length} matching {runInfos.length === 1 ? 'run' : 'runs'}
```

**Review by smurching — COMMENTED** [review]
Generally looks great! There's one weird behavior I noticed during QA where switching experiments will briefly mistakenly show the wrong number of runs while the search runs API request completes for the new experiment, I can show you for debugging purposes

File: mlflow/server/js/src/components/ExperimentPage.js (hunk 11 above)
Comment (Zangr) [reply mlflow/server/js/src/components/ExperimentPage.js]: Added reset in catch.
